### PR TITLE
[TIR] Handle axis_separators during FlattenBuffer

### PR DESCRIPTION
**Pull request by Lunderberg**
For buffers with more than one physical axis, the `axis_separators` are required in order to know which groups of logical axes to fuse into each physical axis.  The implementation in `tir.FlattenBuffer` assumed that all buffers were being flattened to a single physical axis.  Because `tir.LowerOpaqueBlock` replaces the `BlockNode::alloc_buffers` with `Allocate` nodes, `tir.FlattenBuffer` no longer has access to the axis separators and performs inconsistent flattening for `Allocate` as opposed to `BufferLoad`/`BufferStore`.  This was introduced in https://github.com/apache/tvm/pull/12172, which decoupled the lowering/flattening steps.

~~The commit reorders the `tir.FlattenBuffer` to occur before `tir.LowerOpaqueBlock`, to make use of the axis separators.  Any `Allocate` nodes that exist at that point (e.g. from hand-written schedules) are still flattened to 1-d physical buffers, but the `BlockNode::alloc_buffers` are flattened according to the axis separators.~~ See https://github.com/apache/tvm/pull/12652#issuecomment-1233043992.

This PR adds a `DeclBuffer` node to the output of `tir.LowerOpaqueBlock`, which is then used during `tir.FlattenBuffer` to identify the axis separators.

cc @Hzfengsy @junrushao1994

### CODE REVIEW

**Comment by wrongtest-intellif** [comment]
LGTM. Here are my remaining some questions: 
1. Since now we have `decl_buffer` node, if we also convert `T.alloc_buffer` to `T.decl_buffer`  + `T.allocate` in `LowerOpaqueBlock`, could the `axis_separators` get preserved? (I am not so familiar with it, but I notice that it is in the Buffer object's field).  Maybe then the order of the two passes could get totally free.  Our team rely on the IR form which is block-free but multi-dimensional buffer accessing to perform certain analysis and rewriting, thus prefer to lower block before flatten in a customized configuration.

2. Do we have some protection case for the axis_separators feature? It seems #12172 passes without recognizing the introduced issue.

File: tests/python/unittest/test_tir_transform_flatten_buffer.py [hunk 1]
```diff
@@ -20,211 +20,219 @@
...
+    expected = before
+
+
+class TestFlattenInsideBlock(BaseCompare):
+    """Flattening access inside a block flattens the accessed region."""
+
+    def before():
+        A = T.alloc_buffer([32, 32])
+        for i, j in T.grid(32, 32):
+            with T.block("block"):
```
Comment (wrongtest-intellif) [inline tests/python/unittest/test_tir_transform_flatten_buffer.py:203]: Maybe it is better to write explicit read/write annotations?

File: tests/python/unittest/test_tir_transform_flatten_buffer.py [hunk 2]
```diff
@@ -20,211 +20,219 @@
...
+        for i in T.serial(0, 16):
+            B_new = T.alloc_buffer([16], "float32")
+            for j in T.serial(0, 16):
+                B_new[j] = A[((i * 16) + j)] + 1.0
+            for j in T.serial(0, 16):
+                C[((i * 16) + j)] = B_new[j] * 2.0
+
+
+class TestGPU(BaseCompare):
+    """Buffers allocated inside GPU-specific constructs are ignored.
```
Comment (wrongtest-intellif) [inline tests/python/unittest/test_tir_transform_flatten_buffer.py:55]: Does the meaning of "ignored" refer to `B`?

**Comment by Lunderberg** [comment]
> Since now we have decl_buffer node, if we also convert T.alloc_buffer to T.decl_buffer + T.allocate in LowerOpaqueBlock, could the axis_separators get preserved? (I am not so familiar with it, but I notice that it is in the Buffer object's field). Maybe then the order of the two passes could get totally free. Our team rely on the IR form which is block-free but multi-dimensional buffer accessing to perform certain analysis and rewriting, thus prefer to lower block before flatten in a customized configuration.

I like that idea, and having independent order of passes would be better overall.  There was some similar logic in `StorageFlatten` that would look for a BufferLoad/BufferStore in order to know the appropriate axes to flatten, but the `DeclBuffer` usage would be even cleaner.

> Do we have some protection case for the axis_separators feature?

In principle, the `tests/python/contrib/test_hexagon/test_2d_physical_buffers.py::TestElementWise::test_cache_shape` should have caught it, as it validates the shape of a buffer after lowering.  Looking at it again, it uses a TE-based schedule, so it goes through `StorageFlatten` instead of `LowerOpaqueBlock`/`FlattenBuffer`.

Adding a test that validates the buffer shape after running through all of `tvm.lower`, and will be expanding the hexagon-focused PRs in a follow-up.

(Side-note: If `StorageFlatten` were to replace `BufferRealize` with `Allocate` as it currently does, but add a `DeclBuffer` instead of performing the flattening itself, then the same `FlattenBuffer` pass could apply to both types of schedules.  That would be another argument in favor of having `FlattenBuffer` read from `DeclBuffer`, to minimize duplication.)

File: tests/python/unittest/test_tir_transform_flatten_buffer.py [hunk 3]
```diff
@@ -20,211 +20,219 @@
...
+            T.evaluate(A[i, j])
 
+    expected = before
+
+
+class TestFlattenWithAxisSeparators(BaseCompare):
+    """Flattening preserves axis separators"""
+
+    def before():
+        A = T.alloc_buffer([2, 3, 5, 7, 11, 13], axis_separators=[3])
```
[the remark this answers is not in the document]
Comment (Lunderberg) [reply tests/python/unittest/test_tir_transform_flatten_buffer.py:228]: The axis separators are intended to represent memory spaces that should be exposed with multiple indices at the codegen level.  Currently, the only use as part of the hexagon codegen to represent 2-d buffers (`T** buf; T value = buf[i][j]`), but I'd like to also use it to represent GPU texture memory.

For comparison with strided transforms, applying the transform transforming `(i,j) => stride*i + j` would be exposed to the codegen as a single index, where transforming `(i,j) => [i, AXIS_SEPARATOR, j]` would be exposed to to the codegen as two separate values `(i, j)`.

File: tests/python/unittest/test_tir_transform_flatten_buffer.py (hunk 1 above)
Comment (Lunderberg) [reply tests/python/unittest/test_tir_transform_flatten_buffer.py:203]: Good point, especially as that is what this test catches.  Updated.

File: tests/python/unittest/test_tir_transform_flatten_buffer.py (hunk 2 above)
Comment (Lunderberg) [reply tests/python/unittest/test_tir_transform_flatten_buffer.py:55]: Whoops, I was mistaken in that comment, and was getting it mixed up with some of the allocation handling in `StorageRewrite`.  Updated the comment.

**Comment by Lunderberg** [comment]
Inserting the `DeclBuffer` node during `LowerOpaqueBlock`, such that `FlattenBuffer` can be performed after `LowerOpaqueBlock`.  (And renaming the title of this PR accordingly.)

I'm also reverting most of the unit test changes, since they should go back to operating on `Allocate` nodes.  I expect some of these tests to be gated on https://github.com/apache/tvm/pull/12412, which will be necessary to correctly express the axis separators in TVMScript.

**Comment by Lunderberg** [comment]
cc @cconvey

**Comment by Lunderberg** [comment]
@wrongtest-intellif Requesting a re-review, if you have the time, as the implementation changed significantly from the reviewed/approved version.

**Pull request merged**